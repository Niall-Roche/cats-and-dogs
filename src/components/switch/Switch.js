import * as RadixSwitch from '@radix-ui/react-switch'
import {styled} from 'styled-components'
import {CAT_EMOJI, DOG_EMOJI} from '../../constants/images'
import useCatsAndDogs from '../../hooks/useCatsAndDogs'
import {CATS_QUERY, DOGS_QUERY} from '../../constants/staticStrings'

const SwitchRoot = styled(RadixSwitch.Root)`
  cursor: pointer;
  width: 90px;
  height: 46px;
  padding: 0;
  background-color: #b7f4f8;
  border-radius: 100px;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  border: 2px solid #7e7e7e;

  &[data-state='checked'] {
    background-color: #b7f4f8;
  }
`

const SwitchThumb = styled(RadixSwitch.Thumb)`
  display: block;
  width: 37px;
  height: 37px;
  background-color: white;
  border-radius: 9999px;
  transition: transform 150ms;
  transform: translateX(3px);
  border: 1px solid #1e1e1e;

  &[data-state='checked'] {
    transform: translateX(44px);
  }
`

const SwitchImg = styled.img`
  margin-top: 5px;
`

const Switch = () => {
  const {value, setValue} = useCatsAndDogs()

  return (
    <SwitchRoot onCheckedChange={() => setValue(value === CATS_QUERY ? DOGS_QUERY : CATS_QUERY)}>
      <SwitchThumb>
        <SwitchImg height={30} width={30} src={value === CATS_QUERY ? CAT_EMOJI : DOG_EMOJI} />
      </SwitchThumb>
    </SwitchRoot>
  )
}

export default Switch