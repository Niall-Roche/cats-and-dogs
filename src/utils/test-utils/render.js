import {ThemeProvider} from '@tanstack/react-query-devtools/build/lib/theme'
import {render} from '@testing-library/react'
import theme from '../../../app-theme.json'
import {CatsAndDogsProvider} from '../../hooks/useCatsAndDogs'
import {PaginationProvider} from '../../hooks/usePagination'

export default comp => render(
  <CatsAndDogsProvider>
    <PaginationProvider>
      <ThemeProvider theme={theme}>
        {comp}
      </ThemeProvider>
    </PaginationProvider>
  </CatsAndDogsProvider>
)