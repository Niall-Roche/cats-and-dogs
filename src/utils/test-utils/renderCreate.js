import {ThemeProvider} from '@tanstack/react-query-devtools/build/lib/theme'
import TestRenderer from 'react-test-renderer'
import theme from '../../app-theme.json'
import {CatsAndDogsProvider} from '../../hooks/useCatsAndDogs'
import {PaginationProvider} from '../../hooks/usePagination'

export default comp => TestRenderer.create(
  <CatsAndDogsProvider>
    <PaginationProvider>
      <ThemeProvider theme={theme}>
        {comp}
      </ThemeProvider>
    </PaginationProvider>
  </CatsAndDogsProvider>
)