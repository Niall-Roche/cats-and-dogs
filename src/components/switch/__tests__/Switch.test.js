import React from 'react'
import Switch from '../Switch'
import 'jest-styled-components'
import renderCreate from '../../../utils/test-utils/renderCreate'
import {fireEvent, screen} from '@testing-library/react'
import render from '../../../utils/test-utils/render'

describe('Switch', () => {
  it('should match snapshot', () => {
    const comp = renderCreate(<Switch />)
    const tree = comp.toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should match snapshot with required props', () => {
    const comp = renderCreate(
      <Switch
        className='d-flex'
        onChange={() => {}}
      />
    )
    const tree = comp.toJSON()

    expect(tree).toMatchSnapshot()
  })


  it('should toggle the switch', () => {
    const handleToggle = jest.fn()

    render(
      <Switch
        onChange={handleToggle}
      />
    )

    const button = screen.getByRole('switch')

    fireEvent.click(button)

    expect(handleToggle).toHaveBeenCalledTimes(1)
  })
})