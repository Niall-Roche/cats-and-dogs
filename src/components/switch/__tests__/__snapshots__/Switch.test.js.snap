// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should match snapshot 1`] = `
Array [
  <button
    aria-checked={false}
    className="ecZxBy"
    data-state="unchecked"
    data-testid="switch"
    onClick={[Function]}
    role="switch"
    type="button"
    value="on"
  >
    <span
      className="kBgWPn"
      data-state="unchecked"
    >
      <img
        className="drAovU"
        height={30}
        src="cat.png"
        width={30}
      />
      <img
        className="drAovU d-none"
        height={30}
        src="dog.png"
        width={30}
      />
    </span>
  </button>,
  <input
    aria-hidden={true}
    defaultChecked={false}
    style={
      Object {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Switch should match snapshot with required props 1`] = `
Array [
  <button
    aria-checked={false}
    className="ecZxBy d-flex"
    data-state="unchecked"
    data-testid="switch"
    onClick={[Function]}
    role="switch"
    type="button"
    value="on"
  >
    <span
      className="kBgWPn"
      data-state="unchecked"
    >
      <img
        className="drAovU"
        height={30}
        src="cat.png"
        width={30}
      />
      <img
        className="drAovU d-none"
        height={30}
        src="dog.png"
        width={30}
      />
    </span>
  </button>,
  <input
    aria-hidden={true}
    defaultChecked={false}
    style={
      Object {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;
